src/ir/auth_logic/ast.h

```diff
@@ -139,6 +139,46 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class ConstantType { NumberType, PrincipalType };
+
+// Relation declaration have types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class ArgumentType {
+ public:
+  enum class Kind { kNumber, kPrincipal, kCustom };
+  explicit ArgumentType(Kind kind, std::string name)
+      : kind_(std::move(kind)), name_(std::move(name)) {}
+  explicit ArgumentType(Kind kind) : kind_(std::move(kind)) {}
+  const Kind& kind() const { return kind_; }
+  const std::string name() const { return name_; }
+
+ private:
+  Kind kind_;
+  std::string name_;
+};
+
+class RelationDeclaration {
+ public:
+  explicit RelationDeclaration(
+      std::string relation_name, bool is_attribute,
+      std::vector<std::pair<std::string, ArgumentType>> arguments)
```

Comment:
This use of `std::pair` is interesting; should there be a simple `Argument` class that contains within it the fields `std::string argument_name_` and `ArgumentType argument_type_`? It's usually the case that `std::pair` and `std::tuple` should only be used for locals or weird template-y situations. Usually for fields of classes and other more formal layout situations a more formal class should be used.

Reply:
Added a new node for Argument to remove std::pair